File: modules/drag_force_f.m
function [drag_force] = drag_force_f(velocity, helium_mass, altitude)

%************************************************************************
% Purdue Orbital, Flight Dynamics
%
% Project Name: Ascent/Descent Modeling
%
% Function Name: drag_force_f
% File Name: drag_force_f.m
%
% Contributors: Jack Triglianos, Samuel Landers
% Date Created: 11/??/2025
% Last Updated: 11/17/2025
%
% Function Description:
%   Computes the drag force acting on the balloon assuming the balloon
%   behaves as a sphere with a typical drag coefficient. Uses the standard
%   drag equation and atmospheric density as a function of altitude.
%
% References:
%   NASA Glenn Research Center. (n.d.). *Drag of a sphere*. 
%       https://www1.grc.nasa.gov/beginners-guide-to-aeronautics/drag-of-a-sphere/
%   NASA Glenn Research Center. (n.d.). *Drag coefficient*.
%       https://www1.grc.nasa.gov/beginners-guide-to-aeronautics/drag-coefficient/
%
% Input variables:
% - velocity: descent or ascent velocity of balloon, m/s, magnitude varies
% - helium_mass: mass of helium inside balloon, kg, positive
% - altitude: geometric altitude above sea level, m, positive
%
% Output variables:
% - f_d: aerodynamic drag force acting on balloon, N
%
%************************************************************************

% Air density at altitude
air_density = air_density_f(altitude); % kg/m^3

% Balloon cross-sectional area
cross_sectional_area = cross_sectional_area_balloon_f(altitude, helium_mass); % m^2

% Drag coefficient (sphere approximation)
DRAG_COEFFICIENT = 0.47; % dimensionless

% Drag force
drag_force = 0.5 * DRAG_COEFFICIENT * air_density * cross_sectional_area * velocity^2; % N

File: modules/drag_force_f.py
########################################################################
# Purdue Orbital, Flight Dynamics
#
# Project Name: Ascent/Descent Modeling
#
# Function Name: drag_force_f
# File Name: drag_force_f.py
#
# Contributors: Purdue Orbital Flight Dynamics Team
# Date Created: Unknown
# Last Updated: 02/09/2026
#
# Function Description:
#   Computes aerodynamic drag magnitude on the balloon using:
#       D = 0.5 * C_D * rho * A * v^2
#
# Notes:
#   - Returns magnitude (always non-negative). Direction must be applied by
#     the calling code based on velocity sign.
#
# References:
#   None
#
# Input variables:
# - velocity_mps: vertical velocity, m/s, sign varies
# - helium_mass_kg: helium mass in balloon, kg, non-negative
# - altitude_m: geometric altitude, m, non-negative
# - atm: atmosphere dict (SI) from modules.atmosphere.atmosphere_m, must include:
#       - rho_kgm3 (kg/m^3)
#
# Output variables:
# - drag_force_N: drag force magnitude, N, non-negative
#
########################################################################

from __future__ import annotations

from modules.balloon_cross_sectional_area_f import balloon_cross_sectional_area_f

DRAG_COEFF_SPHERE = 0.47  # [-] representative sphere drag coefficient


def drag_force_f(velocity_mps: float, helium_mass_kg: float, altitude_m: float, *, atm: dict) -> float:
    """Return drag magnitude in newtons."""
    air_density_kgm3 = float(atm["rho_kgm3"])  # [kg/m^3]
    area_m2 = balloon_cross_sectional_area_f(altitude_m, helium_mass_kg, atm=atm)  # [m^2]

    drag_force_N = 0.5 * DRAG_COEFF_SPHERE * air_density_kgm3 * area_m2 * (velocity_mps ** 2)  # [N]
    return float(drag_force_N)
